javalanglib/src/main/scala/java/lang/SemanticsUtils.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang
+
+import scala.scalajs.js
+import scala.scalajs.runtime.{linkingInfo, UndefinedBehaviorError}
+
+/** Utilities to test for erroneous conditions depending on the Semantics
+ *  configuration.
+ */
+private[lang] object SemanticsUtils {
+
+  private final val Compliant = 0
+  private final val Fatal = 1
+  private final val Unchecked = 2
+
+  /** Tests for an erroneous condition governed by the `asInstanceOfs`
+   *  semantics.
+   */
+  @inline
+  def asInstanceOfCheck(shouldThrow: js.Function0[scala.Boolean],
+      exception: js.Function0[ClassCastException]): Unit = {
+    genericCheck(linkingInfo.semantics.asInstanceOfs, shouldThrow, exception)
+  }
+
+  /** Tests for an erroneous condition governed by the `arrayIndexOutOfBounds`
+   *  semantics.
+   */
+  @inline
+  def arrayIndexOutOfBoundsCheck(shouldThrow: js.Function0[scala.Boolean],
+      exception: js.Function0[ArrayIndexOutOfBoundsException]): Unit = {
+    genericCheck(linkingInfo.semantics.arrayIndexOutOfBounds, shouldThrow,
+        exception)
+  }
+
+  @inline
+  private def genericCheck(complianceLevel: Int,
+      shouldThrow: js.Function0[scala.Boolean],
+      exception: js.Function0[Throwable]): Unit = {
```

Comment:
Hum, stupid idea: Can we just type this as `js.Function0[RuntimeException]` and get rid of the unwrap exception removal?

Reply:
I thought about that, but it doesn't work, unfortunately, because `JavaScriptException` is also a subclass of `RuntimeException`. :-(

Reply:
... I have an even more stupid idea: Can we simply move the throw at the call site to help the type checker? We can easily guard against forgetting to write `throw` by requiring the return type to be `Nothing`.

Reply:
I tried that, but I could only get it to work by adding a temporary `val`, and in that case the produced code was worse. :(

Reply:
Ah, yes. I didn't notice it is used twice.